server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java

```diff
@@ -388,6 +399,41 @@ public T loadLatestState(Logger logger, NamedXContentRegistry namedXContentRegis
         return null;
     }
 
+
+    /**
+     * Tries to load the latest state from the given data-locations.
+     *
+     * @param logger        a logger instance.
+     * @param dataLocations the data-locations to try.
+     * @return tuple of the latest state and generation. (-1, null) of no state is found.
+     */
+    public Tuple<T, Long> loadLatestStateWithGeneration(Logger logger, NamedXContentRegistry namedXContentRegistry, Path... dataLocations)
+            throws IOException {
+        long generation = findMaxStateId(prefix, dataLocations);
+        T state = loadGeneration(logger, namedXContentRegistry, generation, dataLocations);
+
+        if (generation > -1 && state == null) {
+            throw new IllegalStateException("unable to find state files with state id " + generation +
```

Comment:
Yannick asked me to add this assertion during previous PR review. This could happen if findMaxStateId has found the file, but this file was removed right after it. So it's not an error in program logic (not an AssertionError).